README.md
[![](https://drone.io/daegalus/dart-uuid/status.png)](https://drone.io/daegalus/dart-uuid/latest)

# dart-uuid

Simple, fast generation of [RFC4122](http://www.ietf.org/rfc/rfc4122.txt) UUIDS.

Heavily based on node-uuid by Robert Kieffer (I even copied this readme over and modified it.) 
Primarily becaue it works, well written, and so on.

Features:

* Generate RFC4122 version 1, version 4, or version 5 UUIDs
* Runs in dartvm and hopefully browsers too.
* Cryptographically strong random # generation on all platforms
* [Annotated source code](http://daegalus.github.com/annotated/dart-uuid/Uuid/Uuid.html)

Note: The AES implementation used is a custom written implementation, and I can not guarantee that it actually works. This is only until a standard library version for generating crypt strong random bytes is implemented.

## Getting Started

### Notes on Windows (currently)

* Use cmd for pub
* When actually running an application add // to the beginning of your Package-root directory ([Issue 4534](http://code.google.com/p/dart/issues/detail?id=4534))

### Instructions

1. Install and configure pub (http://www.dartlang.org/docs/pub-package-manager/#installing-and-configuring-pub)
2. Open a command line and cd to your projects root folder
3. In your pubspec, add an entry for dart-uuid to your dependencies (example below)
3. pub install
4. If you wish to run tests, go into packages/dart-uuid/ and run 'dart test/uuid_test.dart'

### Pubspec

There are 2 options. Directly from git, or from pub.dartlang.org

pub.dartlang.org: (you can use 'any' instead of a version if you just want the latest always)
```yaml
dependencies:
  uuid: 0.1.6
```

directly from github:
```yaml
dependencies:
  uuid:
    git: https://github.com/Daegalus/dart-uuid.git
```

```dart
#import('https://raw.github.com/Daegalus/dart-uuid/master/lib/uuid.dart'); //pulls it directly from github.
// OR
#import('package:uuid/uuid.dart'); //Uses the local one in ./packages/uuid/lib/uuid.dart

var uuid = new Uuid();
```

Then create some ids ...

```dart
// Generate a v1 (time-based) id
uuid.v1(); // -> '6c84fb90-12c4-11e1-840d-7b25c5ee775a'

// Generate a v4 (random) id
uuid.v4(); // -> '110ec58a-a0f2-4ac4-8393-c866d813b8d1'

// Generate a v5 (namespace-name-sha1-based) id
uuid.v5(uuid.NAMESPACE_URL, 'www.google.com'); // -> 'c74a196f-f19d-5ea9-bffd-a2742432fc9c'
```

## API

### uuid.v1({Map options: null, List buffer: null, int offset: 0})

Generate and return a RFC4122 v1 (timestamp-based) UUID.

* `options` - (Map) Optional uuid state to apply. Properties may include:

  * `node` - (List) Node id as List of 6 bytes (per 4.1.6). Default: Randomnly generated ID.
  * `clockseq` - (Number between 0 - 0x3fff) RFC clock sequence.  Default: An internally maintained clockseq is used.
  * `msecs` - (Number) Time in milliseconds since unix Epoch.  Default: The current time is used.
  * `nsecs` - (Number between 0-9999) additional time, in 100-nanosecond units. Ignored if `msecs` is unspecified. Default: internal uuid counter is used, as per 4.2.1.2.

* `buffer` - (List) Array or buffer where UUID bytes are to be written.
* `offset` - (Int) Starting index in `buffer` at which to begin writing.

Returns `buffer`, if specified, otherwise the string form of the UUID

Example: Generate string UUID with fully-specified options

```dart
uuid.v1(options: {
    'node': [0x01, 0x23, 0x45, 0x67, 0x89, 0xab],
    'clockSeq': 0x1234,
    'mSecs': new Date(2011,11,01).millisecondsSinceEpoch,
    'nSecs': 5678
})   // -> "1d6a6e2e-0457-11e1-9234-0123456789ab"
```

Example: In-place generation of two binary IDs

```dart
// Generate two ids in an array
var myBuffer = new List(32); // -> []
uuid.v1(buffer: myBuffer);   
// -> [115, 189, 5, 128, 201, 91, 17, 225, 146, 52, 109, 0, 9, 0, 52, 128, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null]
uuid.v1(buffer: myBuffer, offset: 16);  
// -> [115, 189, 5, 128, 201, 91, 17, 225, 146, 52, 109, 0, 9, 0, 52, 128, 115, 189, 5, 129, 201, 91, 17, 225, 146, 52, 109, 0, 9, 0, 52, 128]

// Optionally use uuid.unparse() to get stringify the ids
uuid.unparse(buffer: myBuffer);    // -> '73bd0580-c95b-11e1-9234-6d0009003480'
uuid.unparse(buffer: myBuffer, offset: 16) // -> '73bd0581-c95b-11e1-9234-6d0009003480'
```

### uuid.v4({Map options: null, List buffer: null, int offset: 0})

Generate and return a RFC4122 v4 UUID.

* `options` - (Map) Optional uuid state to apply. Properties may include:

  * `random` - (Number[16]) List of 16 numbers (0-255) to use in place of randomly generated values
  * `rng` - (Function) Random # generator to use. A Custom function that returns an list[16] of byte values or 1 of 2 provided.

* `buffer` - (List) Array or buffer where UUID bytes are to be written.
* `offset` - (Number) Starting index in `buffer` at which to begin writing.

Returns `buffer`, if specified, otherwise the string form of the UUID

Example: Generate string UUID with fully-specified options

```dart
uuid.v4(options: {
  'random': [
    0x10, 0x91, 0x56, 0xbe, 0xc4, 0xfb, 0xc1, 0xea,
    0x71, 0xb4, 0xef, 0xe1, 0x67, 0x1c, 0x58, 0x36
  ]
});
// -> "109156be-c4fb-41ea-b1b4-efe1671c5836"
```

Example: Generate two IDs in a single buffer

```dart
var myBuffer = new List(32);
uuid.v4(buffer: myBuffer);
uuid.v4(buffer: myBuffer, offset: 16);
```

### uuid.v5(String namespace, String name, {Map options: null, List buffer: null, int offset: 0})

Generate and return a RFC4122 v5 UUID.

* `options` - (Map) Optional uuid state to apply. Properties may include:

  * `randomNamespace` - (Boolean) Default True. Returns if you want a v4 generated namespace (true) or NAMESPACE_NIL (false)

* `buffer` - (List) Array or buffer where UUID bytes are to be written.
* `offset` - (Number) Starting index in `buffer` at which to begin writing.

Returns `buffer`, if specified, otherwise the string form of the UUID

Example: Generate string UUID with fully-specified options

```dart
uuid.v5(uuid.NAMESPACE_URL, 'www.google.com');
// -> "c74a196f-f19d-5ea9-bffd-a2742432fc9c"
```

Example: Generate two IDs in a single buffer

```dart
var myBuffer = new List(32);
uuid.v5(uuid.NAMESPACE_URL, 'www.google.com', buffer: myBuffer);
uuid.v5(uuid.NAMESPACE_URL, 'www.google.com', buffer: myBuffer, offset: 16);
```

### uuid.parse(String uuid, {List buffer: null, int offset: 0})
### uuid.unparse(List buffer, {int offset: 0})

Parse and unparse UUIDs

  * `id` - (String) UUID(-like) string
  * `buffer` - (List) Array or buffer where UUID bytes are to be written. Default: A new Array or Buffer is used
  * `offset` - (Int | Number) Starting index in `buffer` at which to begin writing. Default: 0

Example parsing and unparsing a UUID string

```dart
var bytes = uuid.parse('797ff043-11eb-11e1-80d6-510998755d10'); // -> [121, 127, 240, 67, 17, 235, 17, 225, 128, 214, 81, 9, 152, 117, 93, 16]
var string = uuid.unparse(bytes); // -> '797ff043-11eb-11e1-80d6-510998755d10'
```
## Testing

In dartvm

```
dart test\uuid_test.dart
```

In Browser

N/A as I have not used or tested this in the browser.

### Benchmarking

Not ready for this yet. My code is probably inefficient and messy in many areas. Though it should be pretty quick.

## Release notes
See CHANGELOG


CHANGELOG
v0.1.6
- Adjusting usage of constants.
- Fixing tests.

v0.1.5
- Stupid typo on import.

v0.1.4
- Fixing Crypto package move.

v0.1.3
- Fixing language changes.

v0.1.2
- Fix change of charCodes to codeUnits

v0.1.1
- Fixing syntax for upcoming breaking changes.

v0.1.0
- Cleanup, changes, and prep for M3.

v0.0.9
- Minor fix with a const RegExp
- Made sure everything builds on latest dart.
- Fixed pubspec to now import unittest from pub instead of sdk.

v0.0.8
- Changed to the new optional paramater syntaxes and usages.
- Adjusted tests for the new function call style due to parameter change.
- Fixed Import/Source/Library statements to the new format.

v0.0.7
- Made changes requested by the Google team to get my package up on pub.dartlang.org

v0.0.6
- Fixed up some code to make it possibly faster and using better Dart practices.
- Cleaned up some documentation.

v0.0.5
- Added Initial AES for Dart (untested if it actually works/matches other AES encryptors)
- Use AES cipher to create crypto strong bytes.

v0.0.4
- Issue wasn't Math.Random() but a bad reseed by me.
- Cleaned up for new Pub layout.

v0.0.3
- Added UUIDv5
- Fixed UUIDv4 bugs
- Added more unit tests
- Found bug in dart's Math.Random(), reported, waiting for fix to fix my code.

v0.0.2
- Initial tests
- Fixed some parser bugs.

v0.0.1
- Initial Release
- No tests


lib/Uuid.dart
#library('Uuid');
#import('dart:crypto');
#import('dart:math', prefix:"Math");
#import('./aes/aes.dart');

/**
 *  uuid for Dart
 *
 *  Copyright (c) 2012 Yulian Kuncheff
 *
 *  Released under MIT License.
 *
 *  Based on node-uuid by Robert Kieffer.
 */

class Uuid {

  // This isn't used, I just am propogated to use of TAU over PI - http://tauday.com/tau-manifesto
  static final TAU = 2*Math.PI;

  // RFC4122 provided namespaces for v3 and v5 namespace based UUIDs
  const NAMESPACE_DNS = '6ba7b810-9dad-11d1-80b4-00c04fd430c8';
  const NAMESPACE_URL = '6ba7b811-9dad-11d1-80b4-00c04fd430c8';
  const NAMESPACE_OID = '6ba7b812-9dad-11d1-80b4-00c04fd430c8';
  const NAMESPACE_X500= '6ba7b814-9dad-11d1-80b4-00c04fd430c8';
  const NAMESPACE_NIL = '00000000-0000-0000-0000-000000000000';

  var _rndBytes, _seedBytes, _nodeId, _clockSeq, _lastMSecs = 0,
      _lastNSecs = 0;
  var _byteToHex, _hexToByte;

  Uuid() {
    _rndBytes = new List(16);
    _byteToHex = new List(256);
    _hexToByte = new Map();

    // Easy number <-> hex conversion
    for(var i = 0; i < 256; i++) {
      var hex = new List<int>();
      hex.add(i);
      _byteToHex[i] = CryptoUtils.bytesToHex(hex);
      _hexToByte[_byteToHex[i]] = i;
    }

    // Sets initial seedBytes, node, and clock seq based on MathRNG.
    _seedBytes = mathRNG();

    // Per 4.5, create a 48-bit node id (47 random bits + multicast bit = 1)
    _nodeId = [_seedBytes[0] | 0x01,
               _seedBytes[1], _seedBytes[2], _seedBytes[3],
               _seedBytes[4], _seedBytes[5]];

    // Per 4.2.2, randomize (14 bit) clockseq
    _clockSeq = (_seedBytes[6] << 8 | _seedBytes[7]) & 0x3ffff;
  }

  /**
   * Math.Random()-based RNG. All platforms, fast, not cryptographically strong.
   */
  List mathRNG() {
    var rand, b = _rndBytes;

    var _rand = new Math.Random();
    for(var i = 0; i < 16; i++) {
      if ((i & 0x03) == 0) {
        rand = (_rand.nextDouble() * 0x100000000).floor().toInt();
      }
      b[i] = rand >> ((i & 0x03) << 3) & 0xff;
    }

    return b;
  }

  /**
   * AES-based RNG. All platforms, unknown speed, cryptographically strong (theoretically)
   */
  List cryptoRNG() {
    return AES.randomBytes();
  }

  /**
   * Parses the provided [uuid] into a list of byte values.
   * Can optionally be provided a [buffer] to write into and
   *  a positional [offset] for where to start inputting into the buffer.
   */
  List parse(String uuid, [List buffer, int offset=0]) {
    var i = offset, ii = 0;

    // Create a 16 item buffer if one hasn't been provided.
    buffer = (buffer != null) ? buffer : new List(16);

    // Convert to lowercase and replace all hex with bytes then
    // string.replaceAll() does a lot of work that I don't need, and a manual
    // regex gives me more control.
    RegExp regex = const RegExp('[0-9a-f]{2}');
    for(Match match in regex.allMatches(uuid.toLowerCase())) {
      if(ii < 16) {
        var hex = uuid.toLowerCase().substring(match.start(),match.end());
        buffer[i + ii++] = _hexToByte[hex];
      }
    }

    // Zero out any left over bytes if the string was too short.
    while (ii < 16) {
      buffer[i + ii++] = 0;
    }

    return buffer;
  }

  /**
   * Unparses a [buffer] of bytes and outputs a proper UUID string.
   * An optional [offset] is allowed if you want to start at a different point
   *  in the buffer.
   */
  String unparse(List buffer, [int offset=0]) {
    var i = offset;
    return '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}-'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}-'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}-'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}-'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}'
           '${_byteToHex[buffer[i++]]}${_byteToHex[buffer[i++]]}';
  }

  /**
   * v1() Generates a time-based version 1 UUID
   *
   * By default it will generate a string based off current time, and will
   * return a string.
   *
   * If an optional [buffer] list is provided, it will put the byte data into
   * that buffer and return a buffer.
   *
   * Optionally an [offset] can be provided with a start position in the buffer.
   *
   * The first argument is an options map that takes various configuration
   * options detailed in the readme.
   *
   * http://tools.ietf.org/html/rfc4122.html#section-4.2.2
   */
  v1([Map options = null, List buffer, int offset=0]) {
    var i = offset;
    var buf = (buffer != null) ? buffer : new List(16);
    options = (options != null) ? options : new Map();

    var clockSeq = (options['clockSeq'] != null) ? options['clockSeq'] : _clockSeq;

    // UUID timestamps are 100 nano-second units since the Gregorian epoch,
    // (1582-10-15 00:00). Time is handled internally as 'msecs' (integer
    // milliseconds) and 'nsecs' (100-nanoseconds offset from msecs) since unix
    // epoch, 1970-01-01 00:00.
    var mSecs = (options['mSecs'] != null) ? options['mSecs'] : (new Date.now()).millisecondsSinceEpoch;

    // Per 4.2.1.2, use count of uuid's generated during the current clock
    // cycle to simulate higher resolution clock
    var nSecs = (options['nSecs'] != null) ? options['nSecs'] : _lastNSecs + 1;

    // Time since last uuid creation (in msecs)
    var dt = (mSecs - _lastMSecs) + (nSecs - _lastNSecs)/10000;

    // Per 4.2.1.2, Bump clockseq on clock regression
    if (dt < 0 && options['clockSeq'] == null) {
      clockSeq = clockSeq + 1 & 0x3fff;
    }

    // Reset nsecs if clock regresses (new clockseq) or we've moved onto a new
    // time interval
    if ((dt < 0 || mSecs > _lastMSecs) && options['nSecs'] == null) {
      nSecs = 0;
    }

    // Per 4.2.1.2 Throw error if too many uuids are requested
    if (nSecs >= 10000) {
      throw new Exception('uuid.v1(): Can\'t create more than 10M uuids/sec');
    }

    _lastMSecs = mSecs;
    _lastNSecs = nSecs;
    _clockSeq = clockSeq;

    // Per 4.1.4 - Convert from unix epoch to Gregorian epoch
    mSecs += 12219292800000;

    // time Low
    var tl = ((mSecs & 0xfffffff) * 10000 + nSecs) % 0x100000000;
    buf[i++] = tl >> 24 & 0xff;
    buf[i++] = tl >> 16 & 0xff;
    buf[i++] = tl >> 8 & 0xff;
    buf[i++] = tl & 0xff;

    // time mid
    var tmh = (mSecs / 0x100000000 * 10000).toInt() & 0xfffffff;
    buf[i++] = tmh >> 8 & 0xff;
    buf[i++] = tmh & 0xff;

    // time high and version
    buf[i++] = tmh >> 24 & 0xf | 0x10; // include version
    buf[i++] = tmh >> 16 & 0xff;

    // clockSeq high and reserved (Per 4.2.2 - include variant)
    buf[i++] = clockSeq >> 8 | 0x80;

    // clockSeq low
    buf[i++] = clockSeq & 0xff;

    // node
    var node = (options['node'] != null) ? options['node'] : _nodeId;
    for (var n = 0; n < 6; n++) {
      buf[i + n] = node[n];
    }

    return (buffer != null) ? buffer : unparse(buf);
  }

  /**
   * v4() Generates a time-based version 4 UUID
   *
   * By default it will generate a string based AES-based RNG, and will return
   * a string.
   *
   * If an optional [buffer] list is provided, it will put the byte data into
   * that buffer and return a buffer.
   *
   * Optionally an [offset] can be provided with a start position in the buffer.
   *
   * The first argument is an options map that takes various configuration
   * options detailed in the readme.
   *
   * http://tools.ietf.org/html/rfc4122.html#section-4.4
   */
  v4([Map options = null, List buffer, int offset=0]) {
    var i = offset;
    options = (options != null) ? options : new Map();

    // Use the built-in RNG or a custom provided RNG
    var rng = (options['rng'] != null) ? options['rng'] : cryptoRNG();

    // Use provided values over RNG
    var rnds = (options['random'] != null) ? options['random'] : rng;

    // per 4.4, set bits for version and clockSeq high and reserved
    rnds[6] = (rnds[6] & 0x0f) | 0x40;
    rnds[8] = (rnds[8] & 0x3f) | 0x80;

    // Copy the bytes to the buffer if one is provided.
    if (buffer != null) {
      for (var j = 0; j < 16; j++) {
        buffer[i+j] = rnds[j];
      }
    }

    return (buffer != null) ? buffer : unparse(rnds);
  }

  /**
   * v5() Generates a namspace & name-based version 5 UUID
   *
   * By default it will generate a string based on a provided uuid namespace and
   * name, and will return a string.
   *
   * If an optional [buffer] list is provided, it will put the byte data into
   * that buffer and return a buffer.
   *
   * Optionally an [offset] can be provided with a start position in the buffer.
   *
   * The first argument is an options map that takes various configuration
   * options detailed in the readme.
   *
   * http://tools.ietf.org/html/rfc4122.html#section-4.4
   */
  v5(String namespace, String name, [Map options = null, List buffer, int offset = 0]) {
    var i = offset;
    options = (options != null) ? options : new Map();

    // Check if user wants a random namespace generated by v4() or a NIL namespace.
    var useRandom = (options['randomNamespace'] != null) ? options['randomNamespace'] : true;

    // If useRandom is true, generate UUIDv4, else use NIL
    var blankNS = useRandom ? v4() : NAMESPACE_NIL;

    // Use provided namespace, or use whatever is decided by options.
    namespace = (namespace != null) ? namespace : blankNS;

    // Use provided name,
    name = (name != null) ? name : '';

    // Convert namespace UUID to Byte List
    var bytes = parse(namespace);

    // Convert name to a list of bytes
    var nameBytes = new List();
    for(var singleChar in name.charCodes()) {
      nameBytes.add(singleChar);
    }

    // Generate SHA1 using namespace concatenated with name
    var hash = new SHA1();
    hash.update(bytes);
    hash.update(nameBytes);
    List hashBytes = hash.digest();

    // per 4.4, set bits for version and clockSeq high and reserved
    hashBytes[6] = (hashBytes[6] & 0x0f) | 0x50;
    hashBytes[8] = (hashBytes[8] & 0x3f) | 0x80;

    // Copy the bytes to the buffer if one is provided.
    if (buffer != null) {
      for (var j = 0; j < 16; j++) {
        buffer[i+j] = hashBytes[j];
      }
    }

    return (buffer != null) ? buffer : unparse(hashBytes);
  }
}
